src/target.rs

```diff
@@ -0,0 +1,164 @@
+use arrayvec::ArrayVec;
+use std::io::{BufRead, Cursor};
+use std::process::Output;
+use tokio::process::Command;
+
+/// Compiled target triple, used as default for binary fetching
+pub const TARGET: &str = env!("TARGET");
+
+/// Detect the targets supported at runtime,
+/// which might be different from `TARGET` which is detected
+/// at compile-time.
+///
+/// Return targets supported in the order of preference.
+/// If target_os is linux and it support gnu, then it is preferred
+/// to musl.
+///
+/// If target_os is mac and it is aarch64, then aarch64 is preferred
+/// to x86_64.
+///
+/// Check [this issue](https://github.com/ryankurte/cargo-binstall/issues/155)
+/// for more information.
+pub async fn detect_targets() -> ArrayVec<Box<str>, 2> {
```

Comment:
Why use an arrayvec? Performance is unlikely to be a concern at this juncture and it's locking us into at most two potential targets.

Reply:
I mostly want to avoid the heap allocation if possible.

I would post a fix by creating an opaque newtype over it.

Reply:
Done!

With the opaque newtype `Targets`, it can be `Vec`, `ArrayVec` or any other type and we can change its `CAP`.

Reply:
Yeah I'm still doubtful here, avoiding one allocation seems like a pretty thin reason to either use an arrayvec or (now) implement a whole new type, like we've added 5 APIs and a new surface and a bunch of complexity just for that. We're using the vec to store `Box<str>`s (which are allocations) anyway, and this entire thing is going to live for maybe a minute at most and only ever deal with maybe 100 bytes of data, if that, in an application that downloads megabytes of archives and is nearly entirely I/O bound (and in most cases human-input-bound).

Using a standard `Vec` will "just work" and simplify both this code and its use.

Reply:
It's indeed premature optimization and the compiler might optimize it out anyway, so I will revert this to using `Vec`.

Reply:
Done!